src/net/data_stream.rs

```diff
@@ -73,216 +107,143 @@ pub struct AdcDacData {
 /// `stream` is the logically consumer (UDP transmitter) of the enqueued data.
 pub fn setup_streaming(
     stack: NetworkReference,
-) -> (BlockGenerator, DataStream) {
-    let queue = cortex_m::singleton!(: Queue<AdcDacData, BLOCK_BUFFER_SIZE> = Queue::new()).unwrap();
-
+) -> (FrameGenerator, DataStream) {
+    let queue =
+        cortex_m::singleton!(: Queue<StreamFrame, FRAME_COUNT> = Queue::new())
+            .unwrap();
     let (producer, consumer) = queue.split();
 
-    let generator = BlockGenerator::new(producer);
+    let frame_pool =
+        cortex_m::singleton!(: Pool<[u8; FRAME_SIZE]>= Pool::new()).unwrap();
+
+    // Note(unsafe): We guarantee that FRAME_DATA is only accessed once in this function.
+    let memory = unsafe { &mut FRAME_DATA };
+    frame_pool.grow(memory);
+
+    let generator = FrameGenerator::new(producer, frame_pool);
 
-    let stream = DataStream::new(stack, consumer);
+    let stream = DataStream::new(stack, consumer, frame_pool);
 
     (generator, stream)
 }
 
-/// The data generator for a stream.
-pub struct BlockGenerator {
-    queue: Producer<'static, AdcDacData, BLOCK_BUFFER_SIZE>,
-    current_id: u16,
+struct StreamFrame {
+    format: StreamFormat,
+    sequence_number: u16,
+    buffer: Box<[u8; FRAME_SIZE], Init>,
+    offset: usize,
+    batch_count: u16,
+    batch_size: u8,
 }
 
-impl BlockGenerator {
-    /// Construct a new generator.
-    /// # Args
-    /// * `queue` - The producer portion of the SPSC queue to enqueue data into.
-    ///
-    /// # Returns
-    /// The generator to use.
-    fn new(queue: Producer<'static, AdcDacData, BLOCK_BUFFER_SIZE>) -> Self {
+impl StreamFrame {
+    pub fn new(
+        buffer: Box<[u8; FRAME_SIZE], Uninit>,
+        format: StreamFormat,
+        sequence_number: u16,
+    ) -> Self {
         Self {
-            queue,
-            current_id: 0,
+            format,
+            offset: 7,
+            sequence_number,
+            buffer: unsafe { buffer.assume_init() },
```

Comment:
I wonder whether it would be better to follow through with the notion that this is uninitialized data. I would not allow the closure (to `add()`) to read stale/previous data in the buffers.
The closure is the recipient of the "out-pointer" in https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#out-pointers. And the closure will be doing unsafe stuff in general.

Reply:
This is not a `MaybeUninit` structure, but rather a `heapless::pool::Box<T, Uninit>`. The only way to get a `Box<T, Init>` is to call `Box::init(t: T)`, where a copy from `t` -> inner data is performed. If we actually did that, we'd have to enforce a 1KB copy (of zero'd data) into the buffer. I wanted to avoid that overhead, since we never read the buffer anyways.

Reply:
Are you sure. The only example I'm aware of is this: https://github.com/stm32-rs/stm32l4xx-hal/blob/master/examples/rtic_frame_serial_dma.rs#L92
`DMAFrame::new()` has a `MaybeUninit::uninit()` member. The `dma_buf.init()` would copy uninitialized data?

Reply:
`core::ptr::write()` is just a call to memcpy underneath. Thus, even if we have a `MaybeUninit::uninit()` object, we're still copying the entire MaybeUninit structure, which contains the underlying data + some. Check out goldbolt for the compiler output: https://godbolt.org/z/n8r6EexK6 - both do a memcpy of 1024 bytes.

Reply:
Hmm. Right.
Would be interesting to discuss with @korken89 what the intended usage pattern is here without assume_init.